plugin-gen3/auth_impl.go

```diff
@@ -3,97 +3,251 @@ package main
+			field,
+		)
+		return errors.New(msg)
+	}
+	fieldTypeError := func(field string) error {
+		msg := fmt.Sprintf(
+			"failed to decode token: field `%s` has wrong type",
+			field,
+		)
+		return errors.New(msg)
 	}
 
-	pluginConfig := PluginConfig{}
-	err := json.Unmarshal([]byte(jsonConfig), &pluginConfig)
-	if nil != err {
-		return nil, fmt.Errorf("unable to parse JSON configuration: %w", err)
+	jwtApp := authutils.NewJWTApplication("http://fence-service/.well-known/jwks")
+	claims, err := jwtApp.Decode(token)
+	if err != nil {
+		return "", fmt.Errorf("error decoding token: %w", err)
 	}
-	shared.Logger.Info("Configuration", "S3Url", pluginConfig.S3Url)
-	shared.Logger.Info("Configuration", "OidcTokenUrl", pluginConfig.OidcTokenUrl)
-	shared.Logger.Info("Configuration", "OidcClientId", pluginConfig.OidcClientId)
+	scopes := []string{"openid"}
+	expected := &authutils.Expected{Scopes: scopes}
 
-	// exchange the OIDC client ID and secret for an access token
+	err = expected.Validate(claims)
+	if err != nil {
+		return "", fmt.Errorf("error decoding token: %w", err)
+	}
+	userIdInterface, exists := (*claims)["sub"]
+	if !exists {
+		return "", missingRequiredField("sub")
+	}
+	userId, casted := userIdInterface.(string)
+	if !casted {
+		return "", fieldTypeError("sub")
+	}
+
+	return userId, nil
+}
+
+func (a Authorize) PluginAction(params map[string]string, headers map[string]*proto.StringList, configuration *config.Config, task *tes.Task, taskType proto.Type) (*proto.JobResponse, error) {
+	// only proceed for task creation events. The worker config does not need to be updated for
+	// other types of events
+	if taskType == proto.Type_GET || taskType == proto.Type_CANCEL {
+		return &proto.JobResponse{Code: http.StatusOK, Config: configuration, Task: task}, nil
+	}
+	if taskType != proto.Type_CREATE {
+		return &proto.JobResponse{
+				Code:    http.StatusBadRequest,
+				Message: fmt.Sprintf("unsupported task type: %v", taskType),
+			},
+			fmt.Errorf("unsupported task type: %v", taskType)
+	}
+
+	// get the plugin configuration
+	// The OIDC client should be created in Gen3 with:
+	// `fence-create client-create --client CLIENT_NAME --grant-types client_credentials`
+	S3Url, ok := params["S3Url"]
+	if !ok || S3Url == "" {
+		return &proto.JobResponse{
+				Code:    http.StatusBadRequest,
+				Message: "S3Url is required in params",
+			},
+			fmt.Errorf("S3Url is required in params")
+	}
+	OidcClientId, ok := params["OidcClientId"]
+	if !ok || OidcClientId == "" {
+		return &proto.JobResponse{
+				Code:    http.StatusBadRequest,
+				Message: "OidcClientId is required in params",
+			},
+			fmt.Errorf("OidcClientId is required in params")
+	}
+	OidcClientSecret, ok := params["OidcClientSecret"]
+	if !ok || OidcClientSecret == "" {
+		return &proto.JobResponse{
+				Code:    http.StatusBadRequest,
+				Message: "OidcClientSecret is required in params",
+			},
+			fmt.Errorf("OidcClientSecret is required in params")
+	}
```

Comment:
Can we combine this in a loop to avoid code duplication? 
```suggestion

requiredParams := []string{"S3Url", "OidcClientId", "OidcClientSecret"}
paramValues := make(map[string]string)

for _, paramName := range requiredParams {
    value, ok := params[paramName]
    if !ok || value == "" {
        msg := fmt.Sprintf("%s is required in params", paramName)
        return &proto.JobResponse{
                Code:    http.StatusBadRequest,
                Message: msg,
            },
            fmt.Errorf(msg)
    }
    paramValues[paramName] = value
}

S3Url := paramValues["S3Url"]
OidcClientId := paramValues["OidcClientId"]
OidcClientSecret := paramValues["OidcClientSecret"]

```

Reply:
I think this would look even cleaner if we use the `jobResponseError`  helper function I mentioned in the later comments.

Reply:
i think that's maybe a bit overengineered but i'll make the change when i get a chance 😝

Reply:
did not make that change, i think the latest version is good enough, lmk